cloud/amazon/ec2_ami.py

```diff
@@ -354,11 +394,81 @@ def update_image(module, ec2):
     except boto.exception.BotoServerError, e:
         module.fail_json(msg = "%s: %s" % (e.error_code, e.error_message))
 
+def register_image(module, ec2):
+    """
+    Registers AMI
+    """
+
+    name = module.params.get('name')
+    description = module.params.get('description')
+    image_location = module.params.get('image_location')
+    architecture = module.params.get('architecture')
+    kernel_id = module.params.get('kernel_id')
+    root_device_name = module.params.get('root_device_name')
+    virtualization_type = module.params.get('virtualization_type')
+    wait = module.params.get('wait')
+    wait_timeout = int(module.params.get('wait_timeout'))
+    device_mapping = module.params.get('device_mapping')
+
+    try:
+        params = {'name': name,
+                  'description': description,
+                  'architecture': architecture,
+                  'virtualization_type': virtualization_type}
+        if image_location:
+            params['image_location'] = image_location
+        if kernel_id:
+            params['kernel_id'] = kernel_id
+        if root_device_name:
+            params['root_device_name'] = root_device_name
+        
+        if device_mapping:
+            bdm = BlockDeviceMapping()
+            for device in device_mapping:
+                if 'device_name' not in device:
+                    module.fail_json(msg = 'Device name must be set for volume')
+                device_name = device['device_name']
+                del device['device_name']
+                bd = BlockDeviceType(**device)
+                bdm[device_name] = bd
+            params['block_device_map'] = bdm
+
+        image_id = ec2.register_image(**params)
+    except boto.exception.BotoServerError, e:
+        if e.error_code == 'InvalidAMIName.Duplicate':
+            images = ec2.get_all_images()
+            for img in images:
+                if img.name == name:
+                    module.exit_json(msg="AMI name already present", image_id=img.id, state=img.state, changed=False)
+                    sys.exit(0)
+            else:
+                module.fail_json(msg="Error in retrieving duplicate AMI details")
+        else:
+            module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
+
+    # wait here until the image is gone
+    img = ec2.get_image(image_id)
```

Comment:
Am I crazy? Wouldn't this always time out because ec2.get_image() will never return None after successful registration? 
Also, what about an eventual consistency condition when AWS new ami id might not have had propagated, so it's going to throw as non-existent ami id?